samples/read_bam.c

```diff
@@ -0,0 +1,169 @@
+    bam1_t *bamdata = NULL;
+    uint8_t *data = NULL;
+    uint32_t *cigar = NULL;
+
+
+    //read_bam [-n position] [-i infile] [<infile>]
+    while ((c = getopt_long(argc, argv, "i:n:", lopts, NULL)) >= 0)
+    {
+        switch (c)
+        {
+        case 'i':                           //in file name
+            inname = optarg;
+            break;
+        case 'n':                           //position
+            idx = atoi(optarg);
+            break;
+        case '?':                           //unknown options found!
+            print_usage(stderr);
+            goto end;
+            break;
+        }
+    }
+
+    if (argc == 1 && isatty(STDIN_FILENO)) {
+        print_usage(stdout);
+        ret = EXIT_SUCCESS;
+        goto end;
+    }
+
+    if (!inname && ((argc - optind) == 1))
+        inname = argv[optind];
+    if (!inname)
+        inname = "-";
+
+    bamdata = bam_init1();                  //bam data initialization
+    if (!bamdata)
+    {
+        fprintf(stderr, "Failed to allocate data memory!\n");
+        goto end;
+    }
+
+    //open input files
+    infile = sam_open(inname, "r");
+    if (!infile) {
+        fprintf(stderr, "Could not open %s\n", inname);
+        goto end;
+    }
+    //read header
+    in_samhdr = sam_hdr_read(infile);
+    if (!in_samhdr)
+    {
+        fprintf(stderr, "Failed to read header from file!\n");
+        goto end;
+    }
+
+    while (0 <= (ret_r = sam_read1(infile, in_samhdr, bamdata)))
+    {
+        if (idx != cnt++)
+            continue;
+
+        //QNAME FLAG RNAME POS MAPQ CIGAR RNEXT PNEXT TLEN SEQ QUAL [TAG:TYPE:VALUE]…
+        fprintf(stdout, "%s", bam_get_qname(bamdata));                                  //get the query name using the macro
+        fprintf(stdout, "\t%d", bamdata->core.flag);                                    //flag is available in core structure
+        fprintf(stdout, "\t%s", sam_hdr_tid2name(in_samhdr, bamdata->core.tid));        //retrieves the target name using the value in bam and by referring the header
+        fprintf(stdout, "\t%d", bamdata->core.pos + 1);                                 //internally position is 0 based and on text output / SAM it is 1 based
+        fprintf(stdout, "\t%d", bamdata->core.qual);                                    //map quality value
+
+        cigar = bam_get_cigar(bamdata);                                                 //retrieves the cigar data
+        fprintf(stdout, "\t");
+        for (i = 0; i < bamdata->core.n_cigar; ++i) {                                   //no. of cigar data entries
+            fprintf(stdout, "%d%c", bam_cigar_oplen(cigar[i]), bam_cigar_opchr(cigar[i]));  //the macros gives the count of operation and the symbol of operation for given cigar entry
+        }
+        fprintf(stdout, "\t%s", bamdata->core.mtid == bamdata->core.tid ? "=" : sam_hdr_tid2name(in_samhdr, bamdata->core.mtid));   // = if rnext and pnext are same and the target name otherwise
```

Comment:
Good demonstration of the API for querying reference names.  Maybe print it up in both numeric and string form.  Eg:

```
printf("Mate-ref:\t#%d %s\n", b->core.mtid, sam_hdr_tid2name(hdr, b->core.tid));
```

We don't need to try and recreate SAM's "=" notation here either, as we shouldn't attempt to recreate a format ourselves.  We just want to demomstrate how you access the data fields.

Reply:
Updated.